apps/group/contracts/Group.sol

```diff
@@ -45,26 +69,113 @@ contract Group is App, Initializable, IForwarder, EVMCallScriptRunner {
-        members[_entity] = false;
+        isMember[_entity] = false;
+        for (uint i = 0; i < members.length; i++) {
+            if (members[i] == _entity) {
+                members[i] = members[members.length - 1];
+                delete members[members.length - 1];
+                members.length -= 1;
+                break;
+            }
+        }
+
+        if (required > members.length)
+            changeRequirement(members.length);
+
         RemoveMember(_entity);
     }
 
+    function numMembers() public constant returns (uint) {
+        return members.length;
+    }
+
+    /**
+    * @notice Allows a group member to confirm that a specific action can be executed
+    * @param _actionId Index of an action
+    */
+    function confirm(uint _actionId) public {
+        require(isGroupMember(msg.sender));
+        confirmations[_actionId][msg.sender] = true;
+        Confirmation(_actionId, msg.sender);
+    }
+
+    /**
+    * @notice Allows a group member to revoke a confirmation of a specific action
+    * @param _actionId Index of an action
+    */
+    function revoke(uint _actionId) public {
+        require(isGroupMember(msg.sender));
+        confirmations[_actionId][msg.sender] = false;
+        Revocation(_actionId, msg.sender);
+    }
+
+    /**
+    * @notice Determined is a specific action has confirmation to be executed
+    * @param _actionId Index of the action
+    */
+    function isConfirmed(uint _actionId) public constant returns (bool) {
+        uint count = 0;
+        for (uint i = 0; i < members.length; i++) {
+            if (confirmations[_actionId][members[i]]) {
+                count += 1;
+            } if (count == required) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+    * @notice Returns the index of the actionable script
+    * @param _evmCallScript Script being forwarded
+    */
+    function getActionId(bytes _evmCallScript) public constant returns (uint _actionId) {
+        for (uint i = 0; i < actionCount; i++) {
+            if (keccak256(actions[i]) == keccak256(_evmCallScript)) {
+                _actionId = i;
+                break;
+            }
+        }
+    }
+
+    /**
+    * @notice Adds an action for the group to take based on confimations
+    * @param _evmCallScript Script being forwarded
+    */
+    function addAction(bytes _evmCallScript) public constant returns (uint _actionId) {
+        _actionId = actionCount;
+        actions[_actionId] = _evmCallScript;
+        actionCount += 1;
+        ActionAdded(_actionId);
+    }
+
     /**
     * @dev IForwarder interface conformance. Forwards actions to any group member.
     * @param _evmCallScript Script being forwarded
     */
     function forward(bytes _evmCallScript) external {
-        require(isGroupMember(msg.sender));
+        uint actionId = getActionId(_evmCallScript);
+        require(isGroupMember(msg.sender) && isConfirmed(actionId));
```

Comment:
Keep track that the action has been executed so it cannot be executed multiple times.